src/connector/sqlite.rs

```diff
@@ -0,0 +1,285 @@
+use crate::{
+    ast::{Id, ParameterizedValue, Query},
+    error::Error,
+    transaction::{
+        ColumnNames, Connection, Connectional, ResultRow, ToColumnNames, ToResultRow, Transaction,
+        Transactional,
+    },
+    visitor::{self, Visitor},
+    QueryResult, ResultSet,
+};
+use libsqlite3_sys as ffi;
+use r2d2_sqlite::SqliteConnectionManager;
+use rusqlite::{
+    types::{FromSqlError, ValueRef},
+    Connection as SqliteConnection, Row as SqliteRow, Transaction as SqliteTransaction, NO_PARAMS,
+};
+use std::collections::HashSet;
+
+type PooledConnection = r2d2::PooledConnection<r2d2_sqlite::SqliteConnectionManager>;
+type Pool = r2d2::Pool<SqliteConnectionManager>;
+
+pub struct Sqlite {
+    databases_folder_path: String,
+    pool: Pool,
+    test_mode: bool,
+}
+
+impl Transactional for Sqlite {
+    fn with_transaction<F, T>(&self, db: &str, f: F) -> QueryResult<T>
+    where
+        F: FnOnce(&mut Transaction) -> QueryResult<T>,
+    {
+        self.with_connection_internal(db, |ref mut conn| {
+            let mut tx = conn.get_mut().transaction()?;
+            tx.set_prepared_statement_cache_capacity(65536);
+
+            let result = f(&mut tx);
+
+            if result.is_ok() {
+                tx.commit()?;
+            }
+
+            result
+        })
+    }
+}
+
+impl Connectional for Sqlite {
+    fn with_connection<F, T>(&self, db: &str, f: F) -> QueryResult<T>
+    where
+        F: FnOnce(&mut Connection) -> QueryResult<T>,
+    {
+        self.with_connection_internal(db, |c| f(c.get_mut()))
+    }
+}
+
+// Concrete implmentations of trait methods, dropping the mut
+// so we can share it between Connection and Transaction.
+fn execute_impl(conn: &SqliteConnection, q: Query) -> QueryResult<Option<Id>> {
+    let (sql, params) = dbg!(visitor::Sqlite::build(q));
+
+    let mut stmt = conn.prepare_cached(&sql)?;
+    stmt.execute(params)?;
+
+    Ok(Some(Id::Int(conn.last_insert_rowid() as usize)))
+}
+
+fn query_impl(conn: &SqliteConnection, q: Query) -> QueryResult<ResultSet> {
+    let (sql, params) = dbg!(visitor::Sqlite::build(q));
+
+    return query_raw_impl(conn, &sql, &params);
+}
+
+fn query_raw_impl(
+    conn: &SqliteConnection,
+    sql: &str,
+    params: &[ParameterizedValue],
+) -> QueryResult<ResultSet> {
+    let mut stmt = conn.prepare_cached(sql)?;
+    let mut rows = stmt.query(params)?;
+
+    let mut result = ResultSet::empty();
+
+    while let Some(row) = rows.next()? {
+        if result.name_to_index.len() == 0 {
+            result.name_to_index = ResultSet::build_name_map(&row.to_column_names())
```

Comment:
You can do this directly from `Rows` and don't need to call `len` for every row.

https://docs.rs/rusqlite/0.18.0/rusqlite/struct.Rows.html#method.column_names

Reply:
This could also be optional. If we have lots of columns and we don't need the name map, doing this wastes resources.

Reply:
Suggestion: We make loading the names lazy if I manage to pull the iterator/stream implementation of.